File: frontend/src/services/history.ts
import type { LocationProfile } from '../data/locations';
import { roundHalfUp } from '../lib/risk';
import { seeded } from '../lib/rng';
import type { TrendPoint, TrendRange } from '../types/environment';

/** range -> [points, hours between points]; mirrors backend/data/history.py */
export const HISTORY_RANGES: Record<TrendRange, [number, number]> = {
  '24h': [24, 1],
  '7d': [28, 6],
  '30d': [30, 24],
};

const ramp = (current: number, startRatio: number, progress: number) =>
  current * (startRatio + (1 - startRatio) * progress);

export function generateHistory(location: LocationProfile, range: TrendRange, now = new Date()): TrendPoint[] {
  const [points, step] = HISTORY_RANGES[range];
  const rng = seeded(`history:${location.id}:${range}`);
  const { air, water, waste, trend } = location;
  const totalWaste = waste.plastic + waste.paper + waste.other;
  const anchor = new Date(now);
  anchor.setUTCMinutes(0, 0, 0);

  const series: TrendPoint[] = [];
  for (let i = 0; i < points; i += 1) {
    const progress = i / (points - 1);
    const hoursAgo = (points - 1 - i) * step;
    const rAir = rng();
    const rWater = rng();
    const rWaste = rng();
    const latest = i === points - 1;
    const diurnal = Math.sin((2 * Math.PI * hoursAgo) / 24);

    const pm25 =
      air.pm25 === null
        ? null
        : latest
          ? air.pm25
          : roundHalfUp(ramp(air.pm25, trend.air ?? 0.75, progress) * (1 + 0.1 * diurnal) * (1 + (rAir - 0.5) * 0.18), 1);

    const turbidity =
      water.turbidity === null
        ? null
        : latest
          ? water.turbidity
          : roundHalfUp(ramp(water.turbidity, trend.water ?? 0.75, progress) * (1 + (rWater - 0.5) * 0.3), 1);

    const wasteCount = latest
      ? totalWaste
      : roundHalfUp(ramp(totalWaste, trend.waste ?? 0.75, progress) * (1 + (rWaste - 0.5) * 0.4), 0);

    series.push({
      hoursAgo,
      timestamp: new Date(anchor.getTime() - hoursAgo * 3_600_000).toISOString(),
      pm25,
      turbidity,
      wasteCount,
    });
  }
  return series;
}

File: backend/data/history.py
"""Mock historical series for trend charts and anomaly baselines.

Mirrors frontend/src/services/mock/history.ts. Replace with provider history
(OpenAQ /measurements, sensor time-series store, detection log) when live.
"""

import math
from datetime import datetime, timedelta, timezone
from typing import Dict, List, Optional

from core.risk import round_half_up
from core.rng import seeded
from data.locations import Location
from schemas import TrendPoint

# range -> (number of points, hours between points)
RANGES: Dict[str, tuple] = {"24h": (24, 1), "7d": (28, 6), "30d": (30, 24)}


def _ramp(current: float, start_ratio: float, progress: float) -> float:
    return current * (start_ratio + (1 - start_ratio) * progress)


def generate_history(location: Location, range_key: str = "24h") -> List[TrendPoint]:
    if range_key not in RANGES:
        raise ValueError(f"Unsupported range '{range_key}'")

    points, step = RANGES[range_key]
    rng = seeded(f"history:{location['id']}:{range_key}")
    trend = location.get("trend", {})
    air, water, waste = location["air"], location["water"], location["waste"]
    total_waste = waste["plastic"] + waste["paper"] + waste["other"]
    now = datetime.now(timezone.utc).replace(minute=0, second=0, microsecond=0)

    series: List[TrendPoint] = []
    for i in range(points):
        progress = i / (points - 1)
        hours_ago = (points - 1 - i) * step
        r_air, r_water, r_waste = rng(), rng(), rng()
        latest = i == points - 1
        diurnal = math.sin(2 * math.pi * hours_ago / 24)

        pm25: Optional[float] = None
        if air.get("pm25") is not None:
            pm25 = air["pm25"] if latest else round_half_up(
                _ramp(air["pm25"], trend.get("air", 0.75), progress)
                * (1 + 0.1 * diurnal)
                * (1 + (r_air - 0.5) * 0.18),
                1,
            )

        turbidity: Optional[float] = None
        if water.get("turbidity") is not None:
            turbidity = water["turbidity"] if latest else round_half_up(
                _ramp(water["turbidity"], trend.get("water", 0.75), progress)
                * (1 + (r_water - 0.5) * 0.3),
                1,
            )

        waste_count = total_waste if latest else int(
            round_half_up(
                _ramp(total_waste, trend.get("waste", 0.75), progress) * (1 + (r_waste - 0.5) * 0.4),
                0,
            )
        )

        series.append(
            TrendPoint(
                hours_ago=hours_ago,
                timestamp=now - timedelta(hours=hours_ago),
                pm25=pm25,
                turbidity=turbidity,
                waste_count=waste_count,
            )
        )
    return series
